xds/src/main/java/io/grpc/xds/RingHashLoadBalancer.java

```diff
@@ -0,0 +1,373 @@
+    }
+  }
+
+  @Override
+  public void shutdown() {
+    for (Subchannel subchannel : subchannels.values()) {
+      shutdownSubchannel(subchannel);
+    }
+  }
+
+  private void updateBalancingState(SubchannelPicker picker) {
+    checkState(!subchannels.isEmpty(), "no subchannel has been created");
+    ConnectivityState overallState = null;
+    for (Subchannel subchannel : subchannels.values()) {
+      ConnectivityStateInfo stateInfo = getSubchannelStateInfoRef(subchannel).get();
+      overallState = aggregateState(overallState, stateInfo.getState());
+    }
+    if (overallState != currentState || picker != currentPicker || subchannelEnteredReady) {
+      helper.updateBalancingState(overallState, picker);
+      currentState = overallState;
+      currentPicker = picker;
+      subchannelEnteredReady = false;
+    }
+  }
+
+  private static ConnectivityState aggregateState(
+      @Nullable ConnectivityState overallState, ConnectivityState state) {
+    if (overallState == null) {
+      return state;
+    }
+    if (overallState == READY || state == READY) {
+      return READY;
+    }
+    if (overallState == CONNECTING || state == CONNECTING) {
+      return CONNECTING;
+    }
+    if (overallState == IDLE || state == IDLE) {
+      return IDLE;
+    }
+    return overallState;
+  }
+
+  private void processSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateInfo) {
+    if (subchannels.get(stripAttrs(subchannel.getAddresses())) != subchannel) {
+      return;
+    }
+    AtomicReference<ConnectivityStateInfo> subchannelStateRef =
+        getSubchannelStateInfoRef(subchannel);
+
+    // Don't proactively reconnect if the subchannel enters IDLE, even if previously was connected.
+    // If the subchannel was previously in TRANSIENT_FAILURE, it is considered to stay in
+    // TRANSIENT_FAILURE until it becomes READY.
+    if (stateInfo.getState().equals(READY)) {
+      subchannelEnteredReady = true;
+    } else if (subchannelStateRef.get().getState().equals(TRANSIENT_FAILURE)) {
+      return;
+    }
+    subchannelStateRef.set(stateInfo);
+    updateBalancingState(currentPicker);
+  }
+
+  private static AtomicReference<ConnectivityStateInfo> getSubchannelStateInfoRef(
+      Subchannel subchannel) {
+    return checkNotNull(subchannel.getAttributes().get(STATE_INFO), "STATE_INFO");
+  }
+
+  private static final class RingHashPicker extends SubchannelPicker {
+    private final List<RingEntry> ring;
+    // shallow copy of subchannels
+    private final Map<EquivalentAddressGroup, Subchannel> pickableSubchannels;
+
+    private RingHashPicker(
+        List<RingEntry> ring, Map<EquivalentAddressGroup, Subchannel> pickableSubchannels) {
+      this.ring = ring;
+      this.pickableSubchannels = pickableSubchannels;
+    }
+
+    @Override
+    public PickResult pickSubchannel(PickSubchannelArgs args) {
+      long requestHash = args.getCallOptions().getOption(XdsNameResolver.RPC_HASH_KEY);
```

Comment:
The callOption is not set in this class, so better to handle null return value.

Reply:
Fair. Updated.